### PR TITLE
DOC: optimize.least_squares: update `curve_fit` docstring and `ValueError` message

### LINKED ISSUE #20155: DOC: Inconsistent naming convention between docs and error messages within scipy.optimize.curve_fit
### Issue with current documentation:

When working with `scipy.optimize.curve_fit` I had a typo in my bounds, which led to the following ValueError being raised:

```
ValueError: `x0` is infeasible.
```

Referring back to the documentation, the parameter `x0` is not mentioned, with the terminology used throughout the documentation page being `p0` instead: [Documentation](https://scipy.github.io/devdocs/reference/generated/scipy.optimize.curve_fit.html#scipy.optimize.curve_fit)

Looking through the source code, it seems that `p0` is the preferred terminology within the `curve_fit` method, but `x0` is used in many other places in the code. As far as I can see, the message `x0 is infeasible.` only appears here: https://github.com/scipy/scipy/blob/9c4f6a91c5eb3e797d8b591e502d746128dc1ce2/scipy/optimize/_lsq/least_squares.py#L820

### Idea or request for content:

Assuming my understanding of the code is correct, adding a note to the documentation of `curve_fit` that informs a user that `x0` and `p0` are used interchangeably across the codebase would make the user experience a bit more streamlined.

This could be achieved by adding 

> Note that `x0` may be used interchangeably with `p0`

to line 611 of https://github.com/scipy/scipy/blob/main/scipy/optimize/_minpack_py.py

### Additional context (e.g. screenshots, GIFs)

_No response_

### PR DESCRIPTION
**Pull request by m-maggi**
#### Reference issue
Closes #20155.

#### Additional Information
Since I was editing scipy/optimize/_lsq/least_squares.py I also refactored two ".format" strings to f-strings.

### CODE REVIEW

File: scipy/optimize/_minpack_py.py [hunk 1]
```diff
@@ -620,7 +620,7 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
         Initial guess for the parameters (length N). If None, then the
         initial values will all be 1 (if the number of parameters for the
         function can be determined using introspection, otherwise a
-        ValueError is raised).
+        ValueError is raised). Note that `x0` may be used interchangeably with `p0`.
```
Comment (j-bowhay) [inline scipy/optimize/_minpack_py.py:623]: With the change in error message is this statement still needed?

Comment (m-maggi) [reply scipy/optimize/_minpack_py.py:623]: I would remove it. To be on the safe side I committed both a change in the ValueError message and in the documentation. Shall we wait for other opinions?

Comment (melissawm) [reply scipy/optimize/_minpack_py.py:623]: +1 for removing, I think the error message is good now.

**Review by melissawm — APPROVED** [review]

File: scipy/optimize/_minpack_py.py (hunk 1 above)
Comment (mdhaber) [reply scipy/optimize/_minpack_py.py:623]:
```suggestion
        ValueError is raised).
```

File: scipy/optimize/_lsq/least_squares.py [hunk 2]
```diff
@@ -777,8 +777,8 @@ def least_squares(
         raise ValueError("`tr_solver` must be None, 'exact' or 'lsmr'.")
 
     if loss not in IMPLEMENTED_LOSSES and not callable(loss):
-        raise ValueError("`loss` must be one of {} or a callable."
-                         .format(IMPLEMENTED_LOSSES.keys()))
+        raise ValueError(f"`loss` must be one of {IMPLEMENTED_LOSSES.keys()} "
+                           "or a callable.")
```
Comment (mdhaber) [inline scipy/optimize/_lsq/least_squares.py:781]:
```suggestion
        raise ValueError(f"`loss` must be one of {IMPLEMENTED_LOSSES.keys()} "
                         "or a callable.")
```

**Review by mdhaber — CHANGES_REQUESTED** [review]
I resolved the merge conflicts and removed the documentation that is no longer needed. (It actually can be read to mean that the user can pass `x0` as a keyword argument interchangeably with `p0`, which would not be correct.) I also added a test that the error message is produced as intended.

Please either [allow edits from maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests) or merge m-maggi/scipy#1. If I don't hear back within a week or so, I'll go ahead and open a new PR to fix this.

**Comment by m-maggi** [comment]
> I resolved the merge conflicts and removed the documentation that is no longer needed. (It actually can be read to mean that the user can pass `x0` as a keyword argument interchangeably with `p0`, which would not be correct.) I also added a test that the error message is produced as intended.
> 
> Please either [allow edits from maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests) or merge [m-maggi#1](https://github.com/m-maggi/scipy/pull/1). If I don't hear back within a week or so, I'll go ahead and open a new PR to fix this.

Thanks for commenting, I see in this PR that the option "allow edits from maintainers" is ticked (see screenshot below). Should this be ticked in some other place?

Screenshot 2024-07-04 at 16 43 02

**Comment by mdhaber** [comment]
Thanks @m-maggi. I had written in https://github.com/m-maggi/scipy/pull/1#issuecomment-2212292626 that you'd need to remove the protections from the branch or merge my PR, but I went ahead and opened gh-21147 to avoid further delay. You will still be credited as an author.

**Pull request closed without merging**